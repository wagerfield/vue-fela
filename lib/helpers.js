var isArray = require('isarray')
var isObject = require('isobject')
var utils = require('./utils')

var reduceArray = utils.reduceArray
var reduceObject = utils.reduceObject

function mapMethod(map, method, rules, props) {
  if (isArray(map)) {

    return reduceArray(map, function reducer(result, key) {
      result[key] = method(rules[key], props)
      return result
    }, {})

  } else if (isObject(map)) {

    return reduceObject(map, function reducer(result, key, alias) {
      result[alias] = method(rules[key], props)
      return result
    }, {})

  } else {

    return reduceObject(rules, function reducer(result, rule, key) {
      result[key] = method(rule, props)
      return result
    }, {})
  }
}

function getProps(props, defaultProps) {
  return isObject(props) ? props : defaultProps
}

function renderRule(renderer, rule, props) {
  return renderer.renderRule(rule, props)
}

function renderRules(renderer, rules, props, optMap) {
  return mapMethod(optMap, renderer.renderRule, rules, props)
}

function mapRule(rule, optProps) {
  return function className() {
    var props = getProps(optProps, this)
    return renderRule(this.$fela, rule, props)
  }
}

function mapRules(rules, optMap, optProps) {
  return mapMethod(optMap, mapRule, rules, optProps)
}

function mapStyles(rules, optMap, optProps) {
  return function classNames() {
    var props = getProps(optProps, this)
    return renderRules(this.$fela, rules, props, optMap)
  }
}

module.exports = {
  mapRule: mapRule,
  mapRules: mapRules,
  mapStyles: mapStyles,
  renderRule: renderRule,
  renderRules: renderRules
}
