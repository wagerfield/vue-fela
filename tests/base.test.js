import { setup, wrap, snapshot, FelaRenderer } from './helpers'
import Base from './base.vue'

describe('Base component', () => {

  const localVue = setup()

  beforeEach(localVue.prototype.$fela.clear)

  it('has a $fela renderer on the component instance', () => {
    const wrapper = wrap(Base, localVue)
    expect(wrapper.vm.$fela).toEqual(FelaRenderer)
  })

  it('computed prop returns empty string when no rule props value is set', () => {
    const wrapper = wrap(Base, localVue)
    expect(wrapper.vm.one).toBe('')
  })

  it('computed prop returns expected class names when rule props are set', () => {
    const margin = 0
    const color = 'red'
    const w1 = wrap(Base, localVue, { color })
    const w2 = wrap(Base, localVue, { margin })
    const w3 = wrap(Base, localVue, { margin, color })
    expect(w1.vm.one).toBe('a')
    expect(w2.vm.one).toBe('b')
    expect(w3.vm.one).toBe('a b')
  })

  it('renders expected snapshot', () => {
    const margin = 0
    const color = 'red'
    snapshot(wrap(Base, localVue, { color }))
    snapshot(wrap(Base, localVue, { margin }))
    snapshot(wrap(Base, localVue, { margin, color }))
  })
})
