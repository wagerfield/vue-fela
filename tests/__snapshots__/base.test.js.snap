// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base component renders expected snapshot 1`] = `<div class="a"></div>`;

exports[`Base component renders expected snapshot 2`] = `<div class="b"></div>`;

exports[`Base component renders expected snapshot 3`] = `<div class="a b"></div>`;
